File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

public class Solution {
    private static final int INF = 1_000_000_000;

    ArrayList<ArrayList<Node>> map;
    int N, K;

    class Node implements Comparable<Node>{
        int node;
        int dist;

        Node(int node, int dist) {
            this.node = node;
            this.dist = dist;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(this.dist, o.dist);
        }
    }

    public int solution(int N, int[][] road, int K) {
        this.N = N;
        this.K = K;

        map = new ArrayList<>();
        for (int i=0; i<=N; i++) {
            map.add(new ArrayList<>());
        }

        for (int[] r : road) {
            int u = r[0];
            int v = r[1];
            int w = r[2];

            map.get(u).add(new Node(v, w));
            map.get(v).add(new Node(u, w));
        }

        int[] dist = dijkstra();
        int ct = 0;
        for (int i=1; i<=N; i++) {
            if (dist[i] <= K) ct++;
        }

        return ct;
    }

    private int[] dijkstra() {
        int[] dist = new int[N+1];
        Arrays.fill(dist, INF);

        PriorityQueue<Node> pq = new PriorityQueue<>();

        dist[1] = 0;
        pq.offer(new Node(1, 0));

        while (!pq.isEmpty()) {
            Node cur = pq.poll();

            if (cur.dist > dist[cur.node]) {
                continue;
            }

            for (Node nextNode : map.get(cur.node)) {
                int nextDist = cur.dist + nextNode.dist;

                if (nextDist < dist[nextNode.node] && nextDist <= K) {
                    dist[nextNode.node] = nextDist;
                    pq.offer(new Node(nextNode.node, nextDist));
                }
            }
        }

        return dist;
    }
}

File: 프로그래머스/2/12978. 배달/배달.cpp
#include <bits/stdc++.h>

using namespace std;

// 양방향으로 통행할 수 있는 도로로 연결
// 도로별로 걸리는 시간(1~2000)이 다름
// 1번 마을에서 각 마을(1~50개)로 음식 배달을 한다
// 각 마을로부터 K 시간이하로 배달이 가능한 마을에서만 주문을 받는다
// return: 음식 주문을 받을 수 있는 마을의 개수

struct Edge {
    int to, cost;
};

struct Node {
    int node, dist;
    
    bool operator<(const Node& o) const {
        return dist > o.dist; // dist가 더 작은 것이 앞으로
    }
};

vector<int> dijkstra(const vector<vector<Edge>>& map, int n, int K) {
    const int INF = 1e9;
    
    vector<int> dist(n+1, INF);
    
    priority_queue<Node> pq; 
    pq.push({1, 0});
    dist[1] = 0;
    
    while (!pq.empty()) {
        Node cur = pq.top(); pq.pop();
        
        if (cur.dist > dist[cur.node]) {
            continue;
        }
        
        for (const Edge& edge : map[cur.node]) {
            int nextDist = edge.cost + cur.dist;
            if (dist[edge.to] > nextDist && nextDist <= K) {
                dist[edge.to] = nextDist;
                pq.push({edge.to, nextDist});
            } 
        }
    }
    
    return dist;
}

int solution(int N, vector<vector<int> > road, int K) {
    
    vector<vector<Edge>> map(N+1);
    for (const vector<int>& r : road) {
        int u = r[0];
        int v = r[1];
        int w = r[2];
        
        // 양방향 도로
        map[u].push_back({v, w});
        map[v].push_back({u, w});
    }
    
    vector<int> dist = dijkstra(map, N, K);
    
    int ct = 0;
    for (int i=1; i<=N; i++) {
        if (dist[i] <= K) ct++;
    }

    return ct;
}
